From 5352803d6f1f8efdf284b8f1e0495b3e48140188 Mon Sep 17 00:00:00 2001
From: Nikolay Orlyuk <virkony@gmail.com>
Date: Sat, 25 Apr 2015 23:42:39 +0300
Subject: [PATCH] Enhance Lua autoconfiguration

- Detect both luaVERSION and lua-VERSION pkg-config modules
- Allow pkg-config binary overriding
- Keep possibility to select specific LUA_VERSION
---
 build/lua-detect.mk  | 27 ++++++++++++++-------------
 system-autodetect.mk |  3 +++
 2 files changed, 17 insertions(+), 13 deletions(-)

diff --git a/build/lua-detect.mk b/build/lua-detect.mk
index 79b91fd..2b79032 100644
--- a/build/lua-detect.mk
+++ b/build/lua-detect.mk
@@ -13,21 +13,22 @@
 #  * LUA (full path to lua interpreter)
 #  * LUAC (full path to lua compiler)
 
-LUA_VERSION := $(or $(LUA_VERSION), $(shell \
-       (pkg-config --exists lua5.2 && echo 5.2) \
-    || (pkg-config --exists lua5.1 && echo 5.1) \
-    || (pkg-config --exists lua    && echo 5.0)))
+LUA_VERSIONS_CANDIDATES = $(or $(LUA_VERSION),5.2 5.1 5.0)
 
-ifeq ($(LUA_VERSION),)
-    $(error Could not find any lua version. (Did you install the -dev package?))
+LUA_PKG := $(firstword $(foreach ver,$(LUA_VERSIONS_CANDIDATES),$(shell \
+       ($(PKG_CONFIG) --exists lua-$(ver)     && echo lua-$(ver)) \
+    || ($(PKG_CONFIG) --exists lua$(ver:5.0=) && echo lua$(ver:5.0=)))))
+
+ifeq ($(LUA_PKG),)
+    $(error Could not find $(or $(LUA_VERSION),any) lua version. (Did you install the -dev package?))
 endif
 
+LUA_VERSION ?= $(or $(shell $(PKG_CONFIG) --variable=V $(LUA_PKG)),5.0)
+
 # prior to 5.1 the lib didn't include version in name.
-ifeq ($(LUA_VERSION),5.0)
-    LUA_VERSION=
-endif
+LUA_SUFFIX := $(if $(findstring $(LUA_VERSION),5.0),,$(LUA_VERSION))
 
-LUA_LIBS     := $(or $(shell pkg-config --libs lua$(LUA_VERSION)),   $(error "pkg-config couldn't find linker flags for lua$(LUA_VERSION)!"))
-LUA_INCLUDES := $(shell pkg-config --cflags lua$(LUA_VERSION))
-LUA          := $(or $(shell which lua$(LUA_VERSION)),               $(error No lua$(LUA_VERSION) interpreter found!))
-LUAC         := $(or $(shell which luac$(LUA_VERSION)),              $(error No lua$(LUA_VERSION) compiler found!))
+LUA_LIBS     := $(or $(shell $(PKG_CONFIG) --libs $(LUA_PKG)),      $(error "pkg-config couldn't find linker flags for lua$(LUA_SUFFIX)!"))
+LUA_INCLUDES := $(shell $(PKG_CONFIG) --cflags $(LUA_PKG))
+LUA          := $(or $(shell which lua$(LUA_SUFFIX)),               $(error No lua$(LUA_SUFFIX) interpreter found!))
+LUAC         := $(or $(shell which luac$(LUA_SUFFIX)),              $(error No lua$(LUA_SUFFIX) compiler found!))
diff --git a/system-autodetect.mk b/system-autodetect.mk
index 3a9b371..e8abc54 100644
--- a/system-autodetect.mk
+++ b/system-autodetect.mk
@@ -2,6 +2,9 @@
 ## System settings
 ##
 
+# Some system may have custom pkg-config executable name
+PKG_CONFIG ?= pkg-config
+
 ##
 ## Installation paths
 ##
-- 
2.3.3

